### PR TITLE
Change engines handling

### PR DESCRIPTION
**Pull request by lukaszstolarczuk**
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmemkv/245)

### CODE REVIEW

**Comment by lukaszstolarczuk** [comment]
There're still some issues to solve, including failing tests of experimental engines, but please review the idea of adding options for each engine.

**Comment by lukaszstolarczuk** [comment]
one failing test is especially worrying:
[ RUN      ] STreeLargeTest.LargeAscendingTest
[stree] Creating filesystem pool, path=/dev/shm/pmemkv, size=2147483648
unknown file: Failure
C++ exception with description "Failed creating pool" thrown in the test fixture's constructor.
Segmentation fault (core dumped)
Makefile:20: recipe for target 'test' failed

**Comment by lukaszstolarczuk** [comment]
> I think it would be good to have some tests, to check if pmemkv compiles when you disable some engine (to make sure everytihng is under proper ifdef).
> 
so let's say test for each engine enablement and one for all engines disabled (in summary 7 tests)?

> > _[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/245#-:-LfQ-A5R2eHMLEyEeXYK:b6ac01x)_ status: all files reviewed, 1 unresolved discussion (waiting on @lukaszstolarczuk)
> 
> _[CMakeLists.txt, line 151 at r4](https://reviewable.io/reviews/pmem/pmemkv/245#-LfPzMlTD42kKyQ-A976:-LfPzMlTD42kKyQ-A977:b-4bq9ee) ([raw file](https://github.com/pmem/pmemkv/blob/d351ae021090770c75728782a00ff061102033f4/CMakeLists.txt#L151)):_
> 
> > ```cmake
> > if(ENABLE_VSMAP_ENGINE OR ENABLE_VCMAP_ENGINE)
> > 	include(memkind)
> > endif(ENABLE_VSMAP_ENGINE OR ENABLE_VCMAP_ENGINE)
> > ```
> 
> I think you could just write "endif()" here - it would be more readable in my opinion.
endif only in this place? in whole file? In general I don't like any of them, but kept them in all occurrences for consistency (there're some "endif's" I didn't add)

**Comment by lukaszstolarczuk** [comment]
I still see some tests (from experimental engines) failing, but I will report them as bugs.

File: CMakeLists.txt [hunk 1]
```diff
@@ -126,21 +190,26 @@ endif()
 
 add_library(pmemkv SHARED ${SOURCE_FILES})
 
-target_link_libraries(pmemkv PRIVATE ${LIBPMEMOBJ++_LIBRARIES}
-                                    ${MEMKIND_LIBRARIES}
-                                    ${RapidJSON_LIBRARIES}
-                                    ${TBB_LIBRARIES})
+target_link_libraries(pmemkv PRIVATE ${RapidJSON_LIBRARIES})
+if(ENABLE_VSMAP_ENGINE OR ENABLE_VCMAP_ENGINE OR ENABLE_CMAP_ENGINE OR ENABLE_STREE_ENGINE OR ENABLE_TREE3_ENGINE)
+       target_link_libraries(pmemkv PRIVATE ${LIBPMEMOBJ++_LIBRARIES})
```
Comment (marcinslusarz) [inline CMakeLists.txt]: is this really needed? we'll link to obj++ later

File: CMakeLists.txt [hunk 2]
```diff
@@ -126,21 +190,26 @@ endif()
...
+	target_link_libraries(pmemkv PRIVATE ${MEMKIND_LIBRARIES})
+endif()
+if(ENABLE_VCMAP_ENGINE OR ENABLE_CMAP_ENGINE)
+	target_link_libraries(pmemkv PRIVATE ${TBB_LIBRARIES})
+endif()
+if(ENABLE_CACHING_ENGINE)
+	target_link_libraries(pmemkv PRIVATE ${CMAKE_THREAD_LIBS_INIT} memcached)
+	target_link_libraries(pmemkv PRIVATE acl_cpp protocol acl)
+endif()
+target_link_libraries(pmemkv PRIVATE ${LIBPMEMOBJ++_LIBRARIES} ${RapidJSON_LIBRARIES})
```
Comment (marcinslusarz) [inline CMakeLists.txt]: ${RapidJSON_LIBRARIES} is duplicated above

File: src/libpmemkv.cc [hunk 3]
```diff
@@ -30,22 +30,35 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#include <rapidjson/document.h>
+#if defined(ENABLE_VSMAP_ENGINE) || defined(ENABLE_VCMAP_ENGINE) || defined(ENABLE_CMAP_ENGINE) ||\
+    defined(ENABLE_STREE_ENGINE) || defined(ENABLE_TREE3_ENGINE)
 #include <libpmemobj++/make_persistent.hpp>
```
Comment (marcinslusarz) [inline src/libpmemkv.cc]: These includes shouldn't be needed here. Each engine that requires them should include them.

File: utils/docker/run-build.sh [hunk 4]
```diff
@@ -87,3 +87,40 @@ else
...
+do
+	cd $WORKDIR/build
+	rm -rf *
+	# testing each engine separately; disabling default engines
+	cmake .. -DTBB_DIR=/opt/tbb/cmake \
+		-DENABLE_VSMAP_ENGINE=OFF \
+		-DENABLE_VCMAP_ENGINE=OFF \
+		-DENABLE_CMAP_ENGINE=OFF \
+		-D$engine_flag=ON
+	make -j
```
Comment (marcinslusarz) [inline utils/docker/run-build.sh]: max number of jobs is missing

File: utils/docker/run-build.sh [hunk 5]
```diff
@@ -87,3 +87,40 @@ else
...
+# testing all engines turned on
+cd $WORKDIR/build
+rm -rf *
+cmake .. -DTBB_DIR=/opt/tbb/cmake \
+	-DENABLE_VSMAP_ENGINE=ON \
+	-DENABLE_VCMAP_ENGINE=ON \
+	-DENABLE_CMAP_ENGINE=ON \
+	-DENABLE_STREE_ENGINE=ON \
+	-DENABLE_TREE3_ENGINE=ON
+make -j
```
Comment (marcinslusarz) [inline utils/docker/run-build.sh]: .

File: CMakeLists.txt [hunk 6]
```diff
@@ -73,21 +74,79 @@ if(COVERAGE)
...
-    src/engines-experimental/stree/persistent_b_tree.h src/engines-experimental/stree/pstring.h
-    src/engines-experimental/caching.h src/engines-experimental/caching.cc
+# Each engine is switched on separately
+# By default all experimental engines are turned off
+option(ENABLE_VSMAP_ENGINE "enable installation and testing of VSMap engine" ON)
+option(ENABLE_VCMAP_ENGINE "enable installation and testing of VCMap engine" ON)
+option(ENABLE_CMAP_ENGINE "enable installation and testing of CMap engine" ON)
+option(ENABLE_CACHING_ENGINE "enable installation and testing of experimental CachingEngine" OFF)
+option(ENABLE_STREE_ENGINE "enable installation and testing of experimental STree engine" OFF)
+option(ENABLE_TREE3_ENGINE "enable installation and testing of experimental Tree engine" OFF)
```
Comment (marcinslusarz) [inline CMakeLists.txt]: IMHO both names and descriptions are too verbose. 
ENGINE_${name} or ${name}_ENGINE would be enough. ENGINE_${name} is a tiny bit better for automated processing (for example easier to grep/match).
There's no point in saying that option is for "installation and testing" - it should be obvious enough with just "enable [experimental] XXX engine".

**Pull request merged**